### ISSUE TITLE
Catch potential UnboundLocalError in try/except 

### ISSUE THREAD

**Issue by lbenezriravin** [PyCQA/pylint#2835]
Potentially related to #2615. Not sure if these would both be fixed by the same patch, so I'm making a new issue.

### Is your feature request related to a problem? Please describe
The following code passes pylint:
```
try:
   foo = 1 / 0
except:
   pass
print(foo)
```
But raises an `UnboundLocalError` when run. 

### Describe the solution you'd like
Pylint already has an `undefined-loop-variable` warning for a similar case that could result in the same error:
```
for foo in []:
    pass
print(foo)
```
I feel that the former code should raise a warning if that latter code does. More generally, I don't expect `UnboundLocalError` from any code that passes pylint, but I'm not sure if that guarantee is possible.

### Additional context
Thanks for your work on this tool!

**Comment by PCManticore** [comment]
Thanks @lbenezriravin This makes sense.

**Comment by Pierre-Sassoulas** [comment]
Closing as duplicate of #5524

**Comment by DanielNoord** [comment]
This was actually not a duplicate see: https://github.com/PyCQA/pylint/pull/5764#issuecomment-1032803372.

A very relevant comment was made by @cdce8p in #5524 [here](https://github.com/PyCQA/pylint/issues/5524#issuecomment-1025262104):

> Another idea. I like how pyright names it. Maybe `possibly-unbound`?
> 
> ```
> "a" is possibly unbound
> ```
> 
>  > Maybe `consider-using-try-else`
> 
> We shouldn't really "force" users to add an `else` case. There are just to many alternatives. You could for example assign a default value before you enter the `try` block.

**Comment by jacobtylerwalls** [comment]
Implementation tip: when checking whether the except handlers define a name, raise, or return, consider using `NamesConsumer._defines_name_raises_or_returns()` from #5764

**Comment by Pierre-Sassoulas** [comment]
pyright has ``possibly unbound``, mypy has [``possibly-undefined``](https://mypy.readthedocs.io/en/stable/error_code_list2.html#warn-about-variables-that-are-defined-only-in-some-execution-paths-possibly-undefined), I think we should introduce either one of those two names in pylint on top of ``undefined-variable`` (maybe ``possibly-undefined-variable`` for consistency)

**Comment by jacobtylerwalls** [comment]
When I first implemented try/except fathoming in used-before-assignment, the decision we made was to just check the except blocks and let users treat try blocks as succeeding. Even this was a huge change for a minor release, so we stopped there.

This was discussed in a couple places, including here cc/ @DanielNoord:

https://github.com/pylint-dev/pylint/pull/5384#issuecomment-978080181
I'm open to making this change, though, and v4 is probably a good time to get stricter. The implementation is probably not so difficult: see https://github.com/pylint-dev/pylint/issues/2835#issuecomment-1033107462.

**Comment by DanielNoord** [comment]
Seems sensible to me to include this in a `possibly*` message
